File: UI/src/api/monitoring.ts
// ui/src/api/monitoring.ts

import apiClient from './client'
import { useQuery, useQueries } from '@tanstack/react-query'

/* =========================================================
    TYPES & INTERFACES
========================================================= */

export interface MetricAggregatePoint {
  bucket: string
  avg_value: number
  min_value: number
  max_value: number
  sample_count: number
}

export interface TabularResult {
  columns: string[]
  rows: Record<string, any>[]
  collected_at?: string
}

export interface HistoricalDashboardResponse {
  instance: string
  time_range: string
  bucket: string
  timeseries_metrics: Record<string, MetricAggregatePoint[]>
  tabular_metrics: Record<string, TabularResult>
}

export interface CheckRun {
  run_id: number
  started_at: string
  scheduled_at: string
  ended_at: string
  server_id: number
  check_id: number
  status: 'SUCCESS' | 'FAILED' | 'TIMEOUT'
  execution_time_ms: number
  error_message?: string
}

export interface MonitoringLog {
  log_id: number
  collected_at: string
  server_id: number
  check_id: number
  check_name: string
  check_category?: string // Added fallback safely
  raw_result: Record<string, unknown> | any
  result_metadata?: Record<string, unknown> | any
  status_code: 'OK' | 'WARNING' | 'CRITICAL' | 'FAILURE'
  execution_time_ms: number
}

export interface Metric {
  metric_id: number
  collected_at: string
  server_id: number
  check_id?: number
  metric_name: string
  metric_value: number
  labels?: Record<string, unknown>
  details?: Record<string, unknown>
}

export interface RunsSummary {
  total_count: number
  success_count: number
  failed_count: number
  timeout_count: number
  avg_execution_time_ms: number
  success_rate_pct: number | null
  cpu_usage_pct?: number
  memory_usage_pct?: number
  disk_usage_pct?: number
  active_connections?: number
}

export interface RunsAggregatePoint {
  bucket: string
  success_count: number
  failed_count: number
  timeout_count: number
  total_count: number
  success_rate_pct: number | null
  avg_execution_time_ms: number | null
  server_id?: number
  server_label?: string
  check_id?: number
  check_name?: string
  status?: string | number
  started_at?: string
  metric_value?: number | string | null
}

export interface LatestPerCheckRow {
  server_id: number
  server_label: string
  check_id: number
  check_name: string
  check_category: string
  status: string | number
  started_at: string
  execution_time_ms: number | null
  metric_value?: number
  raw_result?: Record<string, unknown> | unknown[]
  result_metadata?: Record<string, unknown> | unknown[] | null
  latest_value?: number | null
}

export interface PaginationMeta {
  total: number
  limit: number
  offset: number
  has_more: boolean
}

export interface ListResponse<T> {
  data: T[]
  meta?: {
    pagination?: PaginationMeta
  }
}

/* =========================================================
    HELPERS
========================================================= */

function normalizeLatestPerCheckRow(
  row: LatestPerCheckRow
): LatestPerCheckRow {
  return {
    ...row,
    metric_value:
      row.metric_value ??
      row.latest_value ??
      (
        Array.isArray(row.result_metadata)
          ? (
              row.result_metadata[0] as Record<string, unknown>
            )?.connection_pct
          : undefined
      ) as number | undefined,
  }
}

/* =========================================================
    HOOKS
========================================================= */

/**
 * Unified query fetching multi-metric time series sets 
 * and deep tabular metrics snapshots via dashboard filters.
 */
export function useHistoricalDashboard(
  serverId?: number,
  timeRange?: '1H' | '6H' | '24H' | '7D',
  bucket?: '5M' | '15M' | '1H' | '6H' | '1D',
  metrics: string[] = []
) {
  return useQuery({
    queryKey: [
      'monitoring',
      'historical',
      'dashboard',
      { serverId, timeRange, bucket, metrics },
    ],

    queryFn: async () => {
      const params = new URLSearchParams()
      
      if (serverId) params.append('server_id', serverId.toString())
      if (timeRange) params.append('time_range', timeRange)
      if (bucket) params.append('bucket', bucket)
      
      metrics.forEach((m) => params.append('metrics', m))

      const response = await apiClient.get<{ data: HistoricalDashboardResponse }>(
        `/monitoring/historical?${params.toString()}`
      )
      return response.data.data
    },

    enabled: !!serverId && !!timeRange && !!bucket && metrics.length > 0,
    placeholderData: (previousData) => previousData,
  })
}

/**
 * FIX: Re-mapped to guarantee MonitoringLog payloads map 
 * straight into DetailDrawer.tsx layout rendering blocks.
 */
export function useCheckRuns(
  serverId?: number,
  checkId?: number,
  from?: string,
  to?: string,
  limit = 50,
  offset = 0
) {
  return useQuery({
    queryKey: [
      'monitoring',
      'logs',
      'list',
      { serverId, checkId, from, to, limit, offset },
    ],

    queryFn: async () => {
      const params = new URLSearchParams()
      if (serverId) params.append('server_id', serverId.toString())
      if (checkId) params.append('check_id', checkId.toString())
      if (from) params.append('from', from)
      if (to) params.append('to', to)
      params.append('limit', limit.toString())
      params.append('offset', offset.toString())

      // Changed from CheckRun to MonitoringLog endpoint to populate metrics
      const response = await apiClient.get<ListResponse<MonitoringLog>>(
        `/monitoring/logs?${params.toString()}`
      )
      return response.data
    },
    enabled: !!serverId && !!checkId,
  })
}

export function useMonitoringLogs(
  serverId?: number,
  checkId?: number,
  from?: string,
  to?: string,
  limit = 50,
  offset = 0
) {
  return useQuery({
    queryKey: [
      'monitoring',
      'logs',
      'details',
      { serverId, checkId, from, to, limit, offset },
    ],

    queryFn: async () => {
      const params = new URLSearchParams()
      if (serverId) params.append('server_id', serverId.toString())
      if (checkId) params.append('check_id', checkId.toString())
      if (from) params.append('from', from)
      if (to) params.append('to', to)
      params.append('limit', limit.toString())
      params.append('offset', offset.toString())

      const response = await apiClient.get<ListResponse<MonitoringLog>>(
        `/monitoring/logs?${params.toString()}`
      )
      return response.data
    },
  })
}

export function useRunsSummary(
  serverId?: number,
  checkId?: number,
  from?: string,
  to?: string
) {
  return useQuery({
    queryKey: [
      'monitoring',
      'logs',
      'summary',
      { serverId, checkId, from, to },
    ],

    queryFn: async () => {
      const params = new URLSearchParams()
      if (serverId) params.append('server_id', serverId.toString())
      if (checkId) params.append('check_id', checkId.toString())
      if (from) params.append('from', from)
      if (to) params.append('to', to)

      const response = await apiClient.get<{ data: RunsSummary }>(
        `/monitoring/logs/summary?${params.toString()}`
      )
      return response.data
    },
    enabled: !!serverId || !!checkId,
  })
}

export function useRunsAggregate(
  serverId?: number,
  checkId?: number,
  from?: string,
  to?: string,
  bucketInterval?: string
) {
  return useQuery({
    queryKey: [
      'monitoring',
      'logs',
      'aggregate',
      { serverId, checkId, from, to, bucketInterval },
    ],

    queryFn: async () => {
      const params = new URLSearchParams()
      if (serverId) params.append('server_id', serverId.toString())
      if (checkId) params.append('check_id', checkId.toString())
      if (from) params.append('from', from)
      if (to) params.append('to', to)
      if (bucketInterval) params.append('bucket_interval', bucketInterval)

      const response = await apiClient.get<{ data: RunsAggregatePoint[] }>(
        `/monitoring/logs/aggregate?${params.toString()}`
      )
      return response.data
    },
    enabled: !!from && !!to,
  })
}

export function useMetrics(
  serverId?: number,
  checkId?: number,
  name?: string,
  from?: string,
  to?: string,
  labels?: string,
  limit = 100,
  offset = 0
) {
  return useQuery({
    queryKey: [
      'monitoring',
      'metrics',
      'list',
      { serverId, checkId, name, from, to, labels, limit, offset },
    ],

    queryFn: async () => {
      const params = new URLSearchParams()
      if (serverId) params.append('server_id', serverId.toString())
      if (checkId) params.append('check_id', checkId.toString())
      if (name) params.append('metric_name', name)
      if (from) params.append('from', from)
      if (to) params.append('to', to)
      if (labels) params.append('labels', labels)
      params.append('limit', limit.toString())
      params.append('offset', offset.toString())

      const response = await apiClient.get<ListResponse<Metric>>(
        `/monitoring/metrics?${params.toString()}`
      )
      return response.data
    },
    enabled: !!serverId,
  })
}

export function useMetricsAggregate(
  serverId: number,
  name?: string,
  bucketInterval?: string,
  from?: string,
  to?: string
) {
  return useQuery({
    queryKey: [
      'monitoring',
      'metrics',
      'aggregate',
      { serverId, name, bucketInterval, from, to },
    ],

    queryFn: async () => {
      const params = new URLSearchParams()
      params.append('server_id', serverId.toString())
      if (name) params.append('metric_name', name)
      if (from) params.append('from', from)
      if (to) params.append('to', to)
      if (bucketInterval) params.append('bucket_interval', bucketInterval)

      const response = await apiClient.get<{ data: MetricAggregatePoint[] }>(
        `/monitoring/metrics/aggregate?${params.toString()}`
      )
      return response.data
    },
    enabled: !!serverId && !!name && !!from && !!to,
  })
}

export function useHistoricalPerCheck(
  serverId?: number,
  checkId?: number,
  from?: string,
  to?: string,
  bucketInterval?: string,
  enabled = false
) {
  const effectiveBucket = bucketInterval || '5m'
  return useQuery({
    queryKey: [
      'monitoring',
      'logs',
      'historical-per-check',
      { serverId, checkId, from, to },
    ],

    queryFn: async () => {
      const params = new URLSearchParams()
      if (serverId) params.append('server_id', serverId.toString())
      if (checkId) params.append('check_id', checkId.toString())
      if (from) params.append('from', from)
      if (to) params.append('to', to)
      params.append('bucket_interval', effectiveBucket)

      const response = await apiClient.get<{ data: any[] }>(
        `/monitoring/logs/historical-per-check?${params.toString()}`
      )
      return response.data
    },
    enabled: enabled && !!from && !!to,
    placeholderData: (previousData) => previousData,
  })
}

export function useLatestPerCheck(
  serverId?: number,
  checkId?: number,
  from?: string,
  to?: string,
  enabled = true
) {
  return useQuery({
    queryKey: [
      'monitoring',
      'logs',
      'latest-per-check',
      { serverId, checkId, from, to },
    ],

    queryFn: async () => {
      const params = new URLSearchParams()
      if (serverId) params.append('server_id', serverId.toString())
      if (checkId) params.append('check_id', checkId.toString())
      if (from) params.append('from', from)
      if (to) params.append('to', to)

      const response = await apiClient.get<{ data: LatestPerCheckRow[] }>(
        `/monitoring/logs/latest-per-check?${params.toString()}`
      )
      return {
        ...response.data,
        data: response.data.data.map(normalizeLatestPerCheckRow),
      }
    },
    enabled,
  })
}

export function useLatestMetrics(
  serverId?: number,
  from?: string,
  to?: string
) {
  return useQuery({
    queryKey: [
      'monitoring',
      'metrics',
      'latest',
      { serverId, from, to },
    ],

    queryFn: async () => {
      const params = new URLSearchParams()
      if (from) params.append('from', from)
      if (to) params.append('to', to)

      const response = await apiClient.get<{ data: Metric[] }>(
        `/monitoring/servers/${serverId}/latest-metrics?${params.toString()}`
      )
      return response.data
    },
    enabled: !!serverId,
  })
}

export function useMetricNames(
  serverId: number,
  from?: string,
  to?: string
) {
  return useQuery({
    queryKey: [
      'monitoring',
      'metrics',
      'names',
      { serverId, from, to },
    ],

    queryFn: async () => {
      const params = new URLSearchParams()
      params.append('server_id', serverId.toString())
      if (from) params.append('from', from)
      if (to) params.append('to', to)

      const response = await apiClient.get<{ data: string[] }>(
        `/monitoring/metrics/names?${params.toString()}`
      )
      return response.data
    },
    enabled: !!serverId,
  })
}

export function useTableCountData(serverIds: number[]) {
  return useQueries({
    queries: serverIds.map(sid => ({
      queryKey: ['monitoring', 'logs', 'latest-per-check', { serverId: sid, checkId: 10 }],
      queryFn: async () => {
        const response = await apiClient.get<{ data: LatestPerCheckRow[] }>(
          `/monitoring/logs/latest-per-check?server_id=${sid}&check_id=10`
        )
        return response.data.data
      },
      staleTime: 60_000,
    })),
    combine: (results) => {
      const allData = results.flatMap(r => r.data ?? [])
      return {
        data: { data: allData },
        isLoading: serverIds.length > 0 && results.some(r => r.isLoading),
        isRefetching: results.some(r => r.isRefetching),
        error: results.find(r => r.error)?.error ?? null,
        refetch: () => Promise.all(results.map(r => r.refetch())),
      }
    },
  })
}

export interface TableCountHistoryPoint {
  collected_at: string
  record_count: number | null
  status: string
}

export function useTableCountHistory(
  serverId: number,
  tableName: string,
  enabled = true
) {
  return useQuery({
    queryKey: ['monitoring', 'table-count-history', { serverId, tableName }],
    queryFn: async () => {
      const params = new URLSearchParams()
      params.append('server_id', serverId.toString())
      params.append('table_name', tableName)
      const response = await apiClient.get<{ data: TableCountHistoryPoint[] }>(
        `/monitoring/logs/table-count-history?${params.toString()}`
      )
      return response.data.data
    },
    enabled: enabled && !!serverId && !!tableName,
  })
}

export interface PartitionCountResult {
  row_count: number
  rows: Record<string, any>[]
}

export function usePartitionCount(
  serverId: number | undefined,
  enabled = true
) {
  return useQuery({
    queryKey: ['monitoring', 'partition-count', { serverId }],
    queryFn: async () => {
      const response = await apiClient.get<{ data: PartitionCountResult }>(
        `/monitoring/partition-count/${serverId}`
      )
      return response.data.data
    },
    enabled: enabled && !!serverId,
    refetchInterval: 60_000,
  })
}

File: API/app/models/responses/monitoring.py
"""
Pydantic response models for monitoring visibility endpoints.

Updated to support deep PostgreSQL checks including tabular data for
Top Queries, Table Bloat, and Index Usage, as well as time-range
and bucketed historical aggregation filters.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Status code → label mappings
# ---------------------------------------------------------------------------

CHECK_RUN_STATUS_MAP: dict[int, str] = {
    1: "SUCCESS",
    2: "FAILED",
    3: "TIMEOUT",
}

MONITORING_LOG_STATUS_MAP: dict[int, str] = {
    1: "WARNING",
    2: "CRITICAL",
    3: "FAILURE",
}

# ---------------------------------------------------------------------------
# Specialized Data Models for DB Internals
# ---------------------------------------------------------------------------

class TabularResult(BaseModel):
    """
    Used for "Top Queries", "Top Tables", "Blocking Sessions", etc.
    Allows the frontend to render dynamic tables.
    """
    columns: List[str] = Field(..., description="List of headers/keys")
    rows: List[Dict[str, Any]] = Field(..., description="The actual data rows")
    collected_at: Optional[datetime] = None  # <--- ADD THIS


class MetricPoint(BaseModel):
    """Simple key-value pair for Gauges and KPI cards."""
    label: str
    value: float
    unit: Optional[str] = None


# ---------------------------------------------------------------------------
# Response models
# ---------------------------------------------------------------------------

class CheckRunResponse(BaseModel):
    run_id: int
    started_at: datetime
    scheduled_at: datetime | None
    ended_at: datetime | None
    server_id: int
    check_id: int
    status: str
    execution_time_ms: int | None
    error_message: str | None


class MonitoringLogResponse(BaseModel):
    log_id: int
    collected_at: datetime
    server_id: int
    check_id: int
    raw_result: Any | None  # Changed to Any to handle lists or dicts
    status_code: str
    execution_time_ms: int | None
    # Added to support tabular visualization in DetailDrawer
    tabular_data: Optional[TabularResult] = None


class MetricResponse(BaseModel):
    metric_id: int
    collected_at: datetime
    server_id: int
    check_id: int
    metric_name: str
    metric_value: float
    labels: dict | None


class MetricAggregatePoint(BaseModel):
    bucket: datetime
    avg_value: float
    min_value: float
    max_value: float
    sample_count: int


class RunsSummaryResponse(BaseModel):
    """
    Aggregated KPIs for the Summary Panel.
    """
    total_count: int = 0
    success_count: int = 0
    failed_count: int = 0
    timeout_count: int = 0
    avg_execution_time_ms: int = 0
    success_rate_pct: float = 0.0

    # Dynamic Live Metrics
    live_metrics: List[MetricPoint] = Field(
        default_factory=list,
        description="Key performance indicators currently active"
    )


class RunsAggregatePoint(BaseModel):
    bucket: datetime
    success_count: int
    failed_count: int
    timeout_count: int
    total_count: int
    success_rate_pct: float | None
    avg_execution_time_ms: int | None


class LatestPerCheckRow(BaseModel):
    server_id: int
    server_label: str
    check_id: int
    check_name: str
    check_category: str
    status: str
    started_at: datetime
    collected_at: Optional[datetime] = None  # <--- ADD THIS
    execution_time_ms: int | None
    latest_value: Optional[float | str] = None
    result_metadata: Optional[Any] = None
    
    # --- CRITICAL FIX: Add this field so the Service/Router doesn't strip it ---
    # This carries the JSON results needed for the Recharts visualizer.
    result_metadata: Optional[Any] = None 


class CheckDetailViewResponse(BaseModel):
    check_id: int
    check_name: str
    last_run_status: str
    # Add this to capture the specific collection time for the drawer header
    collected_at: Optional[datetime] = None  # <--- ADD THIS
    timeseries_data: List[MetricAggregatePoint] = []
    tabular_data: Optional[TabularResult] = None
    summary_stats: List[MetricPoint] = []


# ---------------------------------------------------------------------------
# Filtered Historical Dashboard Response Model
# ---------------------------------------------------------------------------

class HistoricalDashboardResponse(BaseModel):
    """
    Unified payload returning filtered historical snapshots to populate 
    telemetry charts, gauges, and deep DB internal logs.
    """
    instance: str = Field(..., description="Target server/instance label")
    time_range: str = Field(..., description="Selected filter range e.g., 1H, 6H, 24H, 7D")
    bucket: str = Field(..., description="Aggregation group interval e.g., 5M, 15M, 1H, 1D")
    
    # Scalar/Continuous time series mappings (Connections, WAL, Lag, etc.)
    timeseries_metrics: Dict[str, List[MetricAggregatePoint]] = Field(
        default_factory=dict, 
        description="Historical data arrays indexed by metric key name"
    )
    
    # Tabular metrics snapshots corresponding to the selected time context 
    # (Slow Queries, Bloat, Unused Indexes)
    tabular_metrics: Dict[str, TabularResult] = Field(
        default_factory=dict,
        description="Aggregated tabular datasets mapped by check category"
    )


class HistoricalPerCheckPoint(BaseModel):
    """
    A single time-bucketed data point for one of the 9 PostgreSQL checks,
    with fields extracted from the monitoring_logs raw_result JSONB.
    """
    bucket: datetime
    check_id: int

    # Check 1 — Connections
    connection_pct: Optional[float] = None
    total_connections: Optional[int] = None
    active_connections: Optional[int] = None
    idle_connections: Optional[int] = None
    idle_in_txn_connections: Optional[int] = None
    max_connections: Optional[int] = None

    # Check 2 — Blocking Sessions
    blocking_count: Optional[int] = None

    # Check 3 — Table Bloat
    bloat_pct: Optional[float] = None

    # Check 4 — Index Usage
    index_usage_pct: Optional[float] = None

    # Check 5 — Unused Indexes
    unused_count: Optional[int] = None
    total_size: Optional[str] = None

    # Check 6 — Replication Lag
    lag_seconds: Optional[float] = None

    # Check 7 — WAL Usage
    wal_dir_gb: Optional[float] = None
    wal_gb_total: Optional[float] = None
    wal_gb_total_display: Optional[str] = None
    wal_file_count: Optional[int] = None

    # Check 8 — Database Size
    size_gb: Optional[float] = None

    # Check 9 — Slow Queries
    avg_mean_ms: Optional[float] = None

    # Check 10 — Table Count
    table_count: Optional[int] = None

    # Check 11 — Database Age
    xid_age: Optional[int] = None


class TableCountHistoryPoint(BaseModel):
    collected_at: datetime
    record_count: Optional[int] = None
    status: str